tempesta_fw/http.c

```diff
@@ -3873,16 +3873,20 @@ static void
 tfw_http_resp_terminate(TfwHttpMsg *hm)
 {
 	TfwFsmData data;
-	int r;
+	int r = 0;
 
 	/*
-	 * Add absent message framing information. If there is no body, add
-	 *  Content-length header: it requires less modifications.
+	 * Add absent message framing information. It's possible to add a
+	 * 'Content-Length: 0' header, if the Transfer-Encoding header is not
+	 * set, but keep more generic solution and transform to chunked.
+	 * It's the only possible modification for future proxy mode.
+	 * If the framing information can't be added, then close client
+	 * connection after response is forwarded.
 	 */
-	r = (hm->body.len)
-		? tfw_http_msg_to_chunked(hm)
-		: TFW_HTTP_MSG_HDR_XFRM(hm, "Content-Length", "0",
-					TFW_HTTP_HDR_CONTENT_LENGTH, 0);
+	if (!test_bit(TFW_HTTP_B_CHUNKED_APPLIED, hm->flags))
+		r = tfw_http_msg_to_chunked(hm);
```

Comment:
I believe there may be an issue with `tfw_http_msg_to_chunked()` with zero-length body responses. It prepends `<length>\r\n` from the start if the length is non-zero and always appends `\r\n0\r\n\r\n` at the end. If body length is `0`, a body will be replaced with `\r\n0\r\n\r\n` instead of `0\r\n\r\n`. We had no issues with that before since the code called `tfw_http_msg_to_chunked()` only if `hm->body.len > 0`.

Reply:
I have already cared of that case: https://github.com/tempesta-tech/tempesta/pull/1362/files#diff-d2642ed1c8bec1f31e559ba4871eaf51R817
Only `0\r\n\r\n` is added to message with zero body. And I have a functional test for that too: https://github.com/tempesta-tech/tempesta/pull/1362/files#diff-d2642ed1c8bec1f31e559ba4871eaf51R817

Reply:
Yeah, I was looking into the current (master branch) version of `tfw_http_msg_to_chunked()`, and completely forgot that this patch changes `tfw_http_msg_to_chunked()` too.